sdk/core/azure-core-http-vertx/src/main/java/com/azure/core/http/vertx/implementation/VertxRequestWriteSubscriber.java

```diff
@@ -0,0 +1,169 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.http.vertx.implementation;
+
+import com.azure.core.http.HttpResponse;
+import com.azure.core.util.ProgressReporter;
+import com.azure.core.util.logging.ClientLogger;
+import io.netty.buffer.Unpooled;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.core.http.HttpClientRequest;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+import reactor.core.publisher.MonoSink;
+import reactor.core.publisher.Operators;
+import reactor.util.context.Context;
+
+import java.nio.ByteBuffer;
+
+/**
+ * Subscriber that writes a stream of {@link ByteBuffer ByteBuffers} to a {@link HttpClientRequest Vert.x request}.
+ */
+@SuppressWarnings("ReactiveStreamsSubscriberImplementation")
+public final class VertxRequestWriteSubscriber implements Subscriber<ByteBuffer> {
+    private static final ClientLogger LOGGER = new ClientLogger(VertxRequestWriteSubscriber.class);
+
+    private final HttpClientRequest request;
+    private final MonoSink<HttpResponse> emitter;
+    private final ProgressReporter progressReporter;
+
+    // This subscriber is effectively synchronous so there is no need for these fields to be volatile.
+    private volatile Subscription subscription;
+
+    private volatile State state = State.UNINITIALIZED;
+    private volatile Throwable error;
+
+    public VertxRequestWriteSubscriber(HttpClientRequest request, MonoSink<HttpResponse> emitter,
+        ProgressReporter progressReporter) {
+        this.request = request.exceptionHandler(this::onError)
+            .drainHandler(ignored -> requestNext());
+        this.emitter = emitter;
+        this.progressReporter = progressReporter;
+    }
+
+    @Override
+    public void onSubscribe(Subscription s) {
+        // Only set the Subscription if one has not been previously set.
+        // Any additional Subscriptions will be cancelled.
+        if (Operators.validate(subscription, s)) {
+            subscription = s;
+
+            s.request(1);
```

Comment:
Question: is `1` because this subscriber is effectively synchronous and will only be one request to the upstream Publisher?

Reply:
Yeah, this is to keep Reactor from sending too many values too quickly.